gems/smithy-client/lib/smithy-client/param_validator.rb

```diff
@@ -0,0 +1,190 @@
+        # validate_required_union(shape, values, errors)
+
+        # validate non-nil members
+        values.each_pair do |name, value|
+          next if value.nil?
+
+          if shape.member?(name)
+            member_shape = shape.member(name)
+            shape(member_shape.shape, value, errors, context + "[#{name.inspect}]")
+          else
+            errors << "unexpected value at #{context}[#{name.inspect}]"
+          end
+        end
+      end
+
+      def list(shape, values, errors, context)
+        # ensure the value is an array
+        unless values.is_a?(Array)
+          errors << expected_got(context, 'an Array', values)
+          return
+        end
+
+        # validate members
+        values.each.with_index do |value, index|
+          next unless value
+
+          shape(shape.member.shape, value, errors, context + "[#{index}]")
+        end
+      end
+
+      def map(shape, values, errors, context)
+        unless values.is_a?(Hash)
+          errors << expected_got(context, 'a hash', values)
+          return
+        end
+
+        values.each do |key, value|
+          shape(shape.key.shape, key, errors, "#{context} #{key.inspect} key")
+          next unless value
+
+          shape(shape.value.shape, value, errors, context + "[#{key.inspect}]")
+        end
+      end
+
+      # def document(ref, value, errors, context)
+      #   document_types = [Hash, Array, Numeric, String, TrueClass, FalseClass, NilClass]
+      #   unless document_types.any? { |t| value.is_a?(t) }
+      #     errors << expected_got(context, "one of #{document_types.join(', ')}", value)
+      #   end
+      #
+      #   # recursively validate types for aggregated types
+      #   case value
+      #   when Hash
+      #     value.each do |k, v|
+      #       document(ref, v, errors, context + "[#{k}]")
+      #     end
+      #   when Array
+      #     value.each do |v|
+      #       document(ref, v, errors, context)
+      #     end
+      #   end
+      # end
+
+      # rubocop:disable Metrics
+      def shape(shape, value, errors, context)
+        case shape
+        when StructureShape then structure(shape, value, errors, context)
+        when ListShape then list(shape, value, errors, context)
+        when MapShape then map(shape, value, errors, context)
+        # when DocumentShape then document(ref, value, errors, context)
+        when StringShape
+          errors << expected_got(context, 'a String', value) unless value.is_a?(String)
+        when IntegerShape
+          errors << expected_got(context, 'an Integer', value) unless value.is_a?(Integer)
+        when FloatShape
+          errors << expected_got(context, 'a Float', value) unless value.is_a?(Float)
+        when TimestampShape
+          errors << expected_got(context, 'a Time object', value) unless value.is_a?(Time)
+        when BooleanShape
+          errors << expected_got(context, 'true or false', value) unless [true, false].include?(value)
```

Comment:
What about boolish/truthy?

Reply:
But truthy things would not be serialized as true or false?

I think generally the responds_to approach is not correct here but except for blob where we have to check things they can read.